pkg/cmd/release-payload-controller/payload_creation_controller.go

```diff
@@ -0,0 +1,172 @@
+			// If the conditions are both in their respective terminal states, then there is nothing else to do...
+			if (conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) && conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) ||
+				(conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) && conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) {
+				return false
+			}
+			if len(releasePayload.Status.ReleaseCreationJobResult.Status) > 0 {
+				return true
+			}
+		}
+		if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+			if releasePayload, ok := tombstone.Obj.(*v1alpha1.ReleasePayload); ok {
+				// If the conditions are both in their respective terminal states, then there is nothing else to do...
+				if (conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) && conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) ||
+					(conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) && conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) {
+					return false
+				}
+				if len(releasePayload.Status.ReleaseCreationJobResult.Status) >= 0 {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	releasePayloadInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: releasePayloadFilter,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.Enqueue,
+			UpdateFunc: func(old, new interface{}) { c.Enqueue(new) },
+			DeleteFunc: c.Enqueue,
+		},
+	})
+
+	return c, nil
+}
+
+func (c *PayloadCreationController) sync(ctx context.Context, key string) error {
+	klog.V(4).Infof("Starting PayloadCreationController sync")
+	defer klog.V(4).Infof("PayloadCreationController sync done")
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	klog.V(4).Infof("Processing ReleasePayload: '%s/%s' from workQueue", namespace, name)
+
+	// Get the ReleasePayload resource with this namespace/name
+	originalReleasePayload, err := c.releasePayloadLister.ReleasePayloads(namespace).Get(name)
+	// The ReleasePayload resource may no longer exist, in which case we stop processing.
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// If the conditions are both in their respective terminal states, then there is nothing else to do...
+	if (conditions.IsConditionTrue(originalReleasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) && conditions.IsConditionFalse(originalReleasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed)) ||
```

Comment:
don't you actually want: 

if (payloadCreated is true or false) and (payloadFailed is true or false) then we're done.

Reply:
Are they not logically equivalent?  :smile: 
Switched them around.
Done

Reply:
they are not

`(conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated) || conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadCreated)) &&
				(conditions.IsConditionTrue(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed) || conditions.IsConditionFalse(releasePayload.Status.Conditions, v1alpha1.ConditionPayloadFailed))`

Reply:
Updated both blocks to ^^